### PR TITLE
Execute Gradle in parent context in Windows scripts

### PR DESCRIPTION
**Pull request by mataha**
### Context

Windows script `gradlew.bat` executes Gradle within its own, batch context. This causes every locally declared variable to be visible to underlying `java` process, *potentially* leaking information about used options and classpath.

To mitigate this, we jump to a non-existent label, forcing `cmd.exe` to stop processing the script, yet run the remaining commands within the current line in parent (most likely command-line when invoking `gradlew` directly) context. This introduces a harmless side-effect in form of corrupting the current title of the window `cmd.exe` was invoked in, thus it is reset to the default value.

This provides a few side benefits:

  - **no more 'Terminate batch job (Y/N)?' prompt**
  - `call ` and `call` are used to directly set exit code, resolving #4784
  - both sh and `cmd.exe` scripts become synchronized in terms of mechanism used to execute JVM as well as environment sanitization

Issue: #10463

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: gradlew.bat
Comment (ljacomet) [inline gradlew.bat:1]: ❌ This file is generated and should not be modified in a PR

**Review by ljacomet — CHANGES_REQUESTED** [review]
* Can you explain the added deprecations?
* See below for not needed change

Once these are addressed, someone with `bat` knowledge will need to take a look at the proposed changes.

File: gradlew.bat
Comment (mataha) [reply gradlew.bat:1]: Done.

**Comment by mataha** [comment]
> * Can you explain the added deprecations?

Since we're exiting the batch context with `(goto)` and resetting the exit code afterwards, there's no code to run after that line, which makes `exitEnvironmentVar` and its getters/setters obsolete.

I've considered removing it altogether - that said, it would break binary compatibility, so I've chosen to deprecate it instead to show that it's not used nor needed anymore; I'm not 100% set on that though. I just felt that `exitEnvironmentVar` removal should be communicated to the user somehow.

**Comment by ljacomet** [comment]
Thank you for your contribution!

This PR has a valid DCO and meaningful change. The relevant team for this area will confirm the design of the implementation choices.

**Comment by mataha** [comment]
It just occurred to me that this PR fixes #7883 (and its offshoots, like #12237) - after exec-like `(goto)` "succeeds" we're not executing any code in `gradlew.bat` anymore, and so don't have to worry about any negative side-effects of overwriting the file.

Proof-of-concept:

```cmd
@(goto) 2>nul & call & >>"%~dpnx0" (echo(@echo This line wasn't here before...)
```

_Maybe I should update the OP to reflect this..._

File: platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java [hunk 1]
```diff
@@ -157,7 +158,11 @@ private String escapeWindowsJvmOpt(String jvmOpts) {
     private enum ScriptBindingParameter {
         APP_NAME("applicationName"),
         OPTS_ENV_VAR("optsEnvironmentVar"),
+
+        @SuppressWarnings("DeprecatedIsStillUsed")
+        @Deprecated
         EXIT_ENV_VAR("exitEnvironmentVar"),
```
Comment (jvandort) [inline platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java:164]: Why are we leaving this around? I don't see why we need to continue adding a templating binding if we removed the variable from the template

File: platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/scripts/JavaAppStartScriptGenerationDetails.java [hunk 2]
```diff
@@ -37,6 +37,8 @@ public interface JavaAppStartScriptGenerationDetails {
     /**
      * The environment variable to use to control exit value (windows only)
      */
+    @SuppressWarnings("DeprecatedIsStillUsed")
+    @Deprecated
```
Comment (jvandort) [inline platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/scripts/JavaAppStartScriptGenerationDetails.java:41]: If you want to deprecate any public (non-internal) methods, you need to use the `DeprecationLogger` to do so. For this method in particular, you'd need to add a deprecation nag to the implementation in `DefaultJavaAppStartScriptGenerationDetails.java`

It would look something like this:
```
    @Override    
    @Deprecated
    @SuppressWarnings("DeprecatedIsStillUsed")
    public String getExitEnvironmentVar() {
        DeprecationLogger.deprecateMethod(JavaAppStartScriptGenerationDetails.class, "getEnvironmentVar()")
            .willBeRemovedInGradle9()
            .withUpgradeGuideSection(8, "remove_exit_env_var")
            .nagUser();
        
        return exitEnvironmentVar;
    }
```

Any other public deprecated method would need further treatment (deprecated methods in `CreateStartScripts`). 

Next, you need to add an entry to `platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc`. Nothing too complicated, but the entry should be something like below. Make sure the properly link to the javadoc for the properties (my below example does not use the correct syntax for that, but there are examples in the file already).
```
[[remove_exit_env_var]]
==== Deprecate exitEnvironmentVar property in CreateStartScripts task
The `getExitEnvironmentVar` and `setExitEnvironmentVar` methods on `CreateStartScripts` and `getExitEnvironmentVar` methods on `JavaAppStartScriptGenerationDetails` have been deprecated and will be removed in Gradle 9.0. 
```

Assuming we get this PR in for Gradle 8.10, this upgrade guide entry should go under the 8.10 section in the upgrade guide. If it does not exist by the time you update this PR, go ahead and create it, following the pattern of the other gradle versions below it in the file (No need to add empty sections for "potential breaking changes", etc). 

Finally, we need integration tests _verifying_ that these deprecation warnings are emitted. I think it would be fine to just add them to `ApplicationPluginIntegrationTest`, which should already have a `CreateStartScripts` task. Something along the lines of:

```
def "exitEnvironmentVar properties are deprecated"() {
    given:
    buildFile << """
        tasks.startScripts {
            // call the getter and setter methods (all 3 deprecated ones)
        } 
    """
 
    expect:
    executer.expectDocumentedDeprecationWarning(<first warning>)
    executer.expectDocumentedDeprecationWarning(<second warning>)
    executer.expectDocumentedDeprecationWarning(<third warning>)
    succeeds(":startScripts")

}
```

File: platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java [hunk 3]
```diff
@@ -165,9 +165,10 @@ public String getOptsEnvironmentVar() {
     /**
      * The environment variable to use to control exit value (Windows only).
      */
-    @Nullable
-    @Optional
     @Input
+    @Optional
+    @Nullable
+    @Deprecated
```
Comment (jvandort) [inline platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java:171]: Like mentioned in the prior comment, this method (and the below setter) will need deprecation nagging added.

However, since this property is an `@Input`, Gradle will automatically call this method even though there is no reason for it to. This will emit a deprecation warning even if the user does not call the method themselves.

Instead, we should mark this method as `@Internal` so that Gradle does not try to fetch the value when calculating the fingerprint for this task. We can do this since the property value no longer affects task output.

File: gradlew.bat
Comment (mataha) [reply gradlew.bat:1]: Done.

File: platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java (hunk 1 above)
Comment (mataha) [reply platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java]: Removed.

Comment (jvandort) [reply platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java]: Looking back at this, I realized it was possible for users to configure their own templates -- and they still may expect this variable to be present. 

Sorry for the back-and-forth here, but we might need to keep this around.

Comment (mataha) [reply platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java:162]: And it's back.

File: platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/scripts/JavaAppStartScriptGenerationDetails.java (hunk 2 above)
Comment (mataha) [reply platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/scripts/JavaAppStartScriptGenerationDetails.java:41]: Couple questions:

- `exitEnvironmentVar` is a property - shouldn't `deprecateProperty` be used instead?
- Default implementation of `JavaAppStartScriptGenerationDetails` is internal - does it make sense to create integration tests for it?

File: platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java (hunk 3 above)
Comment (mataha) [reply platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java]: Marked as `@Internal`, which in turn forced me to remove `@ToBeReplacedByLazyProperty`.

File: platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java [hunk 4]
```diff
@@ -316,7 +322,13 @@ public void setOptsEnvironmentVar(@Nullable String optsEnvironmentVar) {
         this.optsEnvironmentVar = optsEnvironmentVar;
     }
 
+    @Deprecated
     public void setExitEnvironmentVar(@Nullable String exitEnvironmentVar) {
+        DeprecationLogger.deprecateMethod(CreateStartScripts.class, "setExitEnvironmentVar()")
+            .willBeRemovedInGradle9()
+            .withUpgradeGuideSection(8, "remove_exit_environment_var")
+            .nagUser();
```
Comment (mataha) [inline platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java:330]: Does this have to be marked with `@Internal` as well?

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 5]
```diff
@@ -55,6 +55,20 @@ When this happens, configurations in the substituted project could be selected t
 
 Consuming non-consumable configurations in this manner is deprecated and will result in an error in Gradle 9.0.
 
+[[remove_exit_environment_var]]
+==== Deprecate `exitEnvironmentVar` property in `CreateStartScripts` task
+
+In prior versions of Gradle, Windows Wrapper script ran the main executable within the batch file.
+This could lead to strange behavior during Wrapper upgrades, when Gradle tried to update the script while it was still parsed by `cmd.exe` shell.
+In Gradle 8.10, the Wrapper transfers control back to the parent context in order to invoke the main binary.
+This renders `exitEnvironmentVar` property obsolete as no batch code can run after that point.
```
Comment (mataha) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:64]: I think this looks too convoluted. Hard to phrase it in layman terms though.

**Comment by big-guy** [comment]
We won't be able to look at this again until after 9.0

**Comment by octylFractal** [comment]
Thank you for your contribution, and sorry this wasn't reviewed earlier. After doing some testing on Windows, I found a solution that is more straightforward and doesn't require adjusting the title in #36510. As such, I'm going to close this PR, but I've added you to the contributors list in that PR as I wouldn't have looked into this otherwise.
